bpf/tests/nat_test.h

```diff
@@ -99,3 +102,99 @@ void test_snat_v4_track_local() {
+  *(struct icmphdr *)to = icmp;
+  return 0;
+}
+
+struct __ctx_buff ctx;
+struct ipv4_nat_target target;
+struct ipv4_nat_entry state;
+
+void test_snat_v4_process() {
+
+  resetTest();
+  // Egress
+  // The data in ctx is not valid, revalidate_data returns false.
+  mock_revalidate_data_Stub(revalidate_data_false);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == DROP_INVALID);
+
+  resetTest();
+  // The protocol of iphdr in ctx is not IPPROTO_TCP, IPPROTO_UDP, or
+  // IPPROTO_ICMP.
+  ip4.protocol = IPPROTO_IGMP;
+  mock_revalidate_data_Stub(revalidate_data_true);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == NAT_PUNT_TO_STACK);
+
+  // The protocol of iphdr in ctx is IPPROTO_ICMP and ctx_load_bytes fails and
+  // returns a negative value.
+  ip4.protocol = IPPROTO_ICMP;
+  skb_load_bytes_ExpectAnyArgsAndReturn(-1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == DROP_INVALID);
+
+  // The protocol of iphdr in ctx is IPPROTO_TCP and ctx_load_bytes fails and
+  // returns a negative value.
+  ip4.protocol = IPPROTO_TCP;
+  skb_load_bytes_ExpectAnyArgsAndReturn(-1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == DROP_INVALID);
+
+  // The protocol of iphdr in ctx is IPPROTO_UDP and ctx_load_bytes fails and
+  // returns a negative value.
+  ip4.protocol = IPPROTO_UDP;
+  skb_load_bytes_ExpectAnyArgsAndReturn(-1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == DROP_INVALID);
+
+  // The protocol of iphdr in ctx is IPPROTO_UDP and snat_v4_handle_mapping
+  // returns a negative value.
+  skb_load_bytes_ExpectAnyArgsAndReturn(0);
+  map_lookup_elem_ExpectAnyArgsAndReturn(NULL);
+  mock_ct_lookup4_ExpectAnyArgsAndReturn(-1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, true) == -1);
+
+  // The protocol of iphdr in ctx is IPPROTO_ICMP and the type of icmphdr is
+  // neither ICMP_ECHO nor ICMP_ECHOREPLY.
+  ip4.protocol = IPPROTO_ICMP;
+  icmp.type = ICMP_DEST_UNREACH;
+  skb_load_bytes_Stub(skb_load_bytes_icmphdr);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, false) == DROP_NAT_UNSUPP_PROTO);
+
+  // The protocol of iphdr in ctx is IPPROTO_ICMP and snat_v4_handle_mapping
+  // returns a negative value.
+  ip4.protocol = IPPROTO_ICMP;
+  icmp.type = ICMP_ECHO;
+  map_lookup_elem_ExpectAnyArgsAndReturn(NULL);
+  mock_ct_lookup4_ExpectAnyArgsAndReturn(-1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, true) == -1);
+
+  // The protocol of iphdr in ctx is IPPROTO_ICMP and snat_v4_handle_mapping
+  // returns NAT_CONTINUE_XLATE, then snat_v4_rewrite_egress returns 0.
+  map_lookup_elem_ExpectAnyArgsAndReturn(&state);
+  mock_ct_lookup4_ExpectAnyArgsAndReturn(1);
+  assert(snat_v4_process(&ctx, NAT_DIR_EGRESS, &target, true) == 0);
+
+  // Ingress
+  map_lookup_elem_ExpectAnyArgsAndReturn(&state);
+  mock_ct_lookup4_ExpectAnyArgsAndReturn(1);
+  assert(snat_v4_process(&ctx, NAT_DIR_INGRESS, &target, true) == 0);
```

Comment:
Maybe an additional negative INGRESS test would also be nice to verify that path.

Reply:
I didn't block the PR though to wait for it. Up to you.

Reply:
Thanks for the comment. INGRESS diverges from EGRESS in the last line, so I think it does not matter if we add an additional negative INGRESS test or not?